### PR TITLE
[Refact](type system) refact serde for type system and pb serde impl

### PR DESCRIPTION
**Pull request by amorynan**
# Proposed changes

Issue Number: close #xxx

## Problem summary

Describe your changes.

## Checklist(Required)

* [ ] Does it affect the original behavior
* [ ] Has unit tests been added
* [ ] Has document been added or modified
* [ ] Does it need to update dependencies
* [ ] Is this PR support rollback (If NO, please explain WHY)

## Further comments

If this is a relatively large or complex change, kick off the discussion at [dev@doris.apache.org](mailto:dev@doris.apache.org) by explaining why you chose the solution you did and what alternatives you considered, etc...

### CODE REVIEW

File: be/src/vec/data_types/data_type_nullable.h [hunk 1]
```diff
@@ -98,6 +99,10 @@ class DataTypeNullable final : public IDataType {
     const DataTypePtr& get_nested_type() const { return nested_data_type; }
     bool is_null_literal() const override { return nested_data_type->is_null_literal(); }
 
+    DataTypeSerDeSPtr get_serde() const override {
+        return std::make_shared<DataTypeNullableSerDe>(nested_data_type->get_serde());
```
Comment (yiguolei) [inline be/src/vec/data_types/data_type_nullable.h:103]: nullable 可能不对，nullable 可能是有2列，一列是nested，另外一列是表示一列是否为null，比如下面的代码：
void RPCFnImpl::convert_nullable_col_to_pvalue(const ColumnPtr& column,
                                               const DataTypePtr& data_type,
                                               const ColumnUInt8& null_col, PValues* arg, int start,
                                               int end) {
    int row_count = end - start;
    if (column->has_null(row_count)) {
        auto* null_map = arg->mutable_null_map();
        null_map->Reserve(row_count);
        const auto* col = check_and_get_column<ColumnUInt8>(null_col);
        auto& data = col->get_data();
        null_map->Add(data.begin() + start, data.begin() + end);
        RPCFnImpl::convert_col_to_pvalue<true>(column, data_type, arg, start, end);
    } else {
        RPCFnImpl::convert_col_to_pvalue<false>(column, data_type, arg, start, end);
    }
}

template <bool nullable>
void RPCFnImpl::convert_col_to_pvalue(const ColumnPtr& column, const DataTypePtr& data_type,
                                      PValues* arg, int start, int end) {

File: be/src/vec/data_types/data_type_struct.h [hunk 2]
```diff
@@ -99,6 +100,14 @@ class DataTypeStruct final : public IDataType {
     std::string to_string(const IColumn& column, size_t row_num) const override;
     void to_string(const IColumn& column, size_t row_num, BufferWritable& ostr) const override;
     bool get_have_explicit_names() const { return have_explicit_names; }
+    DataTypeSerDeSPtr get_serde() const override {
+        DataTypeSerDeSPtrs ptrs;
+        ptrs.reserve(elems.size());
```
Comment (yiguolei) [inline be/src/vec/data_types/data_type_struct.h:105]: 这种代码实际没啥用。。。 这里不是很关键，不会有性能问题的

File: be/src/vec/data_types/serde/data_type_bitmap_serde.cpp [hunk 3]
```diff
@@ -0,0 +1,40 @@
...
+// under the License.
+
+#include "data_type_bitmap_serde.h"
+
+namespace doris {
+
+namespace vectorized {
+Status DataTypeBitMapSerDe::write_column_to_pb(const IColumn& column, PValues& result, int start,
+                                               int end) const {
+    result.mutable_bytes_value()->Reserve(end - start);
```
Comment (yiguolei) [inline be/src/vec/data_types/serde/data_type_bitmap_serde.cpp:25]: ptype->set_id(PGenericType::BITMAP);  is lost?

File: be/src/vec/data_types/serde/data_type_bitmap_serde.cpp [hunk 4]
```diff
@@ -0,0 +1,40 @@
...
+    for (size_t row_num = start; row_num < end; ++row_num) {
+        StringRef data = column.get_data_at(row_num);
+        result.add_bytes_value(data.to_string());
+    }
```
Comment (yiguolei) [inline be/src/vec/data_types/serde/data_type_bitmap_serde.cpp:29]: we should consider the nullable 
    case TypeIndex::BitMap: {
        ptype->set_id(PGenericType::BITMAP);
        arg->mutable_bytes_value()->Reserve(row_count);
        for (size_t row_num = start; row_num < end; ++row_num) {
            if constexpr (nullable) {
                if (column->is_null_at(row_num)) {
                    arg->add_bytes_value(nullptr);
                } else {
                    StringRef data = column->get_data_at(row_num);
                    arg->add_bytes_value(data.data, data.size);
                }
            } else {
                StringRef data = column->get_data_at(row_num);
                arg->add_bytes_value(data.data, data.size);
            }
        }
        break;
    }

File: be/src/vec/data_types/serde/data_type_decimal_serde.h [hunk 5]
```diff
@@ -0,0 +1,64 @@
...
+    Status write_column_to_pb(const IColumn& column, PValues& result, int start, int end) const override;
+    Status read_column_from_pb(IColumn& column, const PValues& arg) const override;
+};
+
+template <typename T>
+Status DataTypeDecimalSerDe<T>::write_column_to_pb(const IColumn& column, PValues& result,
+                                                   int start, int end) const {
+    int row_count = end - start;
+    const auto* col = check_and_get_column<ColumnDecimal<T>>(column);
+    if constexpr (std::is_same_v<T, Decimal<Int128>>) {
```
Comment (yiguolei) [inline be/src/vec/data_types/serde/data_type_decimal_serde.h:39]: forget decimal32,64,128i

File: be/src/vec/data_types/serde/data_type_hll_serde.cpp [hunk 6]
```diff
@@ -0,0 +1,41 @@
...
+namespace doris {
+
+namespace vectorized {
+
+Status DataTypeHLLSerDe::write_column_to_pb(const IColumn& column, PValues& result, int start,
+                                            int end) const {
+    result.mutable_bytes_value()->Reserve(end - start);
+    for (size_t row_num = start; row_num < end; ++row_num) {
+        StringRef data = column.get_data_at(row_num);
+        result.add_bytes_value(data.to_string());
```
Comment (yiguolei) [inline be/src/vec/data_types/serde/data_type_hll_serde.cpp:28]: performance maybe bad.
add_bytes(data.c_str, data.size)?

File: be/src/vec/functions/function_rpc.cpp [hunk 7]
```diff
@@ -316,6 +317,7 @@ void RPCFnImpl::convert_col_to_pvalue(const ColumnPtr& column, const DataTypePtr
     }
 }
 
+// (Todo.Amory) need to delete here
 template <bool nullable>
```
Comment (yiguolei) [inline be/src/vec/functions/function_rpc.cpp:321]: 这个 nullable 模板参数是没用的

**Comment by amorynan** [comment]
run buildall

**Comment by hello-stephen** [comment]
TeamCity pipeline, clickbench performance test result:
 the sum of best hot time: 34.17 seconds
 stream load tsv:          434 seconds loaded 74807831229 Bytes, about 164 MB/s
 stream load json:         21 seconds loaded 2358488459 Bytes, about 107 MB/s
 stream load orc:          59 seconds loaded 1101869774 Bytes, about 17 MB/s
 stream load parquet:          33 seconds loaded 861443392 Bytes, about 24 MB/s
 https://doris-community-test-1308700295.cos.ap-hongkong.myqcloud.com/tmp/20230417033454_clickbench_pr_129844.html

**Comment by amorynan** [comment]
run p0

**Comment by amorynan** [comment]
run buildall

**Review by yiguolei — DISMISSED** [review]
LGTM

**Comment by amorynan** [comment]
run buildall

**Review by yiguolei — APPROVED** [review]
lgtm

**Pull request merged**